pageserver/src/layered_repository.rs

```diff
@@ -158,6 +158,13 @@ pub struct LayeredRepository {
     // Global pageserver config parameters
     pub conf: &'static PageServerConf,
 
+    // Allows us to gracefully cancel operations that edit the directory
+    // that backs this layered repository. Usage:
+    //
+    // Use `let _guard = file_lock.try_read()` while writing any files.
+    // Use `let _guard = file_lock.write().unwrap()` to wait for all writes to finish.
+    pub file_lock: RwLock<()>,
```

Comment:
With ithis wording we actually need to take this lock for walreceiver too, because it flushes data to disk with its separate thread

Reply:
I guess also for branch creation and checkpoint. Seems like an issue for a separate PR.

Reply:
The flush is a checkpoint basically. Branch creation holds gc lock for that reason :) 

Lets mention here that it does not cover all situations with a todo or something